Issue 414: list constraints

jeswr opened the pull request:
Closes #414. This PR implements `sh:minListLength`, `sh:maxListLength` and `sh:uniqList`.

<!--
# Pull request instructions
(Note that text between the HTML comments will be recorded in this pull request's source for future editing, but will not render on Github.  Text is suggested with portions needing substitution enclosed in curly braces.  The curly braces also need to be removed when substituting.)

Please use "Closing keywords" for the associated Issue if this PR should close the Issue once merged.
https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

If this PR edits a document, please consider including a "Live render."  The non-comment template text suggests one mechanism.  If including the link line(s), please test-load the link(s) while drafting the PR.
-->

* [See this document rendered online here](https://raw.githack.com/w3c/data-shapes/feat/list-constraints/shacl12-core/index.html)

---


My interpretation of the following from the definition of [2.1.1 Constraints, Parameters and Constraint Components](https://raw.githack.com/w3c/data-shapes/feat/memberShape/shacl12-core/index.html#constraints) is that because the `sh:MemberShapeConstraintComponent` now has multiple parameters; there can only be one occurrence of `sh:memberShape` in a given shape - and I have updated the `sh:maxCount` accordingly.

> Some constraint components declare only a single parameter. For example [sh:ClassConstraintComponent](https://raw.githack.com/w3c/data-shapes/feat/memberShape/shacl12-core/index.html#ClassConstraintComponent) has the single parameter sh:class. These parameters may be used multiple times in the same shape, and each [value](https://raw.githack.com/w3c/data-shapes/feat/memberShape/shacl12-core/index.html#dfn-values) of such a parameter declares an individual [constraint](https://raw.githack.com/w3c/data-shapes/feat/memberShape/shacl12-core/index.html#dfn-constraints). The interpretation of such declarations is conjunction, i.e. all constraints apply. The following example specifies that the values of ex:customer have to be [SHACL instances](https://raw.githack.com/w3c/data-shapes/feat/memberShape/shacl12-core/index.html#dfn-shacl-instance) of both ex:Customer and ex:Person.

---

⚠️⚠️ ⚠️ 
This is currently targeted to #403 branch for a smaller diff in the GH UI.  #403 should be merged and this targeted to `gh-pages` rather than merging this branch into #403

jeswr committed:
#401 fix: constrain deactivated using datatype rather than in (#402)

jeswr committed:
feat: add memberShape to vocab, shacl-shacl and tests

jeswr committed:
chore: add HTML documentation

jeswr committed:
fix: use SHACL list rather than RDF list, allow memberShape to be used on Node Shapes and add appropriate tests

jeswr committed:
Revert "#401 fix: constrain deactivated using datatype rather than in (#402)" (#404)

This reverts commit 3a1573a1bdd664580c9cf0257346cf3ec45cba3a.

jeswr committed:
Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

jeswr committed:
Update shacl12-core/index.html

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

jeswr committed:
chore: cleanup newlines

jeswr committed:
chore: fixup ordering of constraint components

jeswr committed:
chore: remove maxCount constraint on memberShape

jeswr committed:
chore: replace list of constraint parameters with reference

jeswr committed:
chore: fix memberShape validation example in text and 001 test

jeswr committed:
chore: cleanup tests

jeswr committed:
chore: update vocab definition

jeswr committed:
chore: add memberShape test for multiple violating values

jeswr committed:
chore: cleanup turtle

jeswr committed:
cleanup: remove unecessary otherwise

jeswr committed:
Update shacl12-vocabularies/shacl.ttl

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

jeswr committed:
chore: add clarifying comments

jeswr committed:
chore: add clarifications to tests

jeswr committed:
feat: use sh:detail

jeswr committed:
chore: update membershape text definition

jeswr committed:
chore: fix must casing

jeswr committed:
chore: add better sh:detail examples for invalid lists

jeswr committed:
chore: correctly reference _:b1 in property memberShape

jeswr committed:
chore: fix formatting of memberShape test

jeswr committed:
chore: add link to sh:detail results for sh:memberShape

jeswr committed:
chore: move link to sh:detail results for sh:memberShape

jeswr committed:
chore: remove maxCount form memberShape vocab definition

jeswr committed:
chore: add vocab definitions for sh:minListLength, sh:maxListLength, and sh:uniqList

jeswr committed:
chore: add list constraints

jeswr committed:
chore: mark sh:minListLength, sh:maxListLength and sh:uniqList as optional

jeswr committed:
chore: add list tests

jeswr committed:
chore: cleanup sh:memberShape text deifnition

jeswr committed:
chore: cleanup sh:memberShape text deifnition

jeswr committed:
fix list test url

HolgerKnublauch reviewed (changes requested):
I think you have linked the new constraint parameters to the same constraint component for memberShape. I believe it would be better to introduce separate constraint components for each of them. It makes the definition more modular but, more importantly, it is perfectly sensible to use for example sh:minListLength without a sh:memberShape constraint.

HolgerKnublauch commented on shacl12-core/index.html:
```diff
@@ -5715,21 +5715,59 @@ <h4>sh:memberShape</h4>
 						<div class="def-text-body" data-validator="MemberShape">
 							Let <code>$memberShape</code> be a <a>value</a> of <code>sh:memberShape</code>.
 							Each <a>value node</a> <code>v</code> must be a <a href="#syntax-rule-SHACL-list">SHACL list</a> - a <a>failure</a> MUST be produced if <code>v</code> is not a SHACL list.
+							For each <a>value node</a> <code>v</code> that is a valid SHACL list, a <a>failure</a> MUST be produced for each violation of the following conditions:
```
Note that FAILURE is not the right word here. It should be "validation result". Failure is used when a node shape fails to evaluate due to some low-level problem. Normal constraint violations are validation results.

jeswr replied:
I have fixed this, aligned the textual definition with the other textual definitions in the document and moved discussion of `sh:detail` to the informative section in doing so.

HolgerKnublauch commented on shacl12-core/index.html:
```diff
@@ -5715,21 +5715,59 @@ <h4>sh:memberShape</h4>
+									If <code>sh:maxListLength</code> is present, the number of members in the list <code>v</code> MUST be less than or equal to the value of <code>sh:maxListLength</code>.
+								</li>
+								<li>
+									If <code>sh:uniqList</code> is present and has the value <code>true</code>, the members of the list <code>v</code> MUST be unique.
```
I still don't like the name uniqList, and would prefer uniqueMembers. The abbreviation uniq is not really necessary for such a short word, and not the list is unique but its members are.

jeswr commented:
> I think you have linked the new constraint parameters to the same constraint component for memberShape. I believe it would be better to introduce separate constraint components for each of them. It makes the definition more modular but, more importantly, it is perfectly sensible to use for example sh:minListLength without a sh:memberShape constraint.

Makes sense! Following this approach; I expect similarly to `sh:memberShape` - there should be an error if `sh:minListLength` is declared and the focus node is not a valid SHACL list. This would mean that if both `sh:memberShape` and `sh:minListLength` are declared - and the focus node is not a valid list then both constraints would independently trigger validation results which look like https://github.com/w3c/data-shapes/blob/536ce96fe3c6f1a1c774f569292a281d198aac8d/shacl12-test-suite/tests/core/node/memberShape-001.ttl#L121-L138

I guess we can expect similar behaviour if there are multiple `sh:memberShape` declarations in #403

Is this the behaviour you would expect for validation results?

HolgerKnublauch commented:
I wouldn't worry too much about malformed lists. Usually people use tools to produce them, including editing TTL source code. So if they are reported twice, this shouldn't be the end of the world.

jeswr commented:
I also saw a comment about lists not being valid objects of `sh:targetNode` but it's now not appearing in the GH UI - was this comment deleted because the statement doesn't hold, or is GH doing weird things?

HolgerKnublauch commented:
Yeah my fault. I deleted the comment because I understood later what you did.

jeswr committed:
chore: update vocab, shacl and tests to make sh:minListLength, sh:maxListLength, sh:uniqueMembers separate constraint components

jeswr committed:
chore: remove non-existent files from manifest

jeswr committed:
chore: separate list validators separate components in the spec document

jeswr committed:
chore: add examples to list constraints

jeswr committed:
chore: fix unique members namespace

jeswr commented:
I think this is good to go now - I suggest we merge this to `gh-pages` and close #403 as I have applied changes to the `sh:memberShape` definition in this PR including moving it to the new [List Constraint Component](https://raw.githack.com/w3c/data-shapes/feat/list-constraints/shacl12-core/index.html#core-components-list) section and addressing https://github.com/w3c/data-shapes/pull/416#discussion_r2173705517

HolgerKnublauch commented on shacl12-vocabularies/shacl-shacl.ttl:
```diff
@@ -226,7 +227,24 @@ shsh:ShapeShape
+		sh:minInclusive 0 ;             # maxListLength-minInclusive
+	] ;
+	sh:property [
+		sh:path sh:uniqList ;
```
sh:uniqList -> sh:uniqueMembers

jeswr replied:
```suggestion
		sh:path sh:uniqueMembers ;
```

HolgerKnublauch commented on shacl12-core/index.html:
```diff
@@ -4517,6 +4520,462 @@ <h4>sh:uniqueLang</h4>
+					<p><em>The remainder of this section is informative.</em></p>
+					<p>
+						Each member <code>m</code> of a value node <code>v</code> that does not conform to the <code>$memberShape</code> should be reported as a separate <code>sh:detail</code> in the <a>validation result</a> for <code>v</code>.
+						If the list <code>v</code> is not a valid <a href="#syntax-rule-SHACL-list">SHACL list</a>, this should also be reported as a <code>sh:detail</code> in the <a>validation result</a> for <code>v</code>.
```
I think this needs to be changed. If the list is not valid, it should become a separate top-level violation and it shouldn't even attempt to validate the individual members anymore. This applies to all List-based constraint components IMHO, including length.

jeswr replied:
Changed in https://github.com/w3c/data-shapes/pull/416/commits/ca91c2c71ce6526995d39a5ba2eca744ba2d9d39. Length based constraints were already top-level violations so I don't believe any changes are needed to those tests - unless there is an expectation that there are distinct values for some of the result properties.

HolgerKnublauch commented on shacl12-core/index.html:
```diff
@@ -4517,6 +4520,462 @@ <h4>sh:uniqueLang</h4>
+							Let <code>$uniqueMembers</code> be a <a>parameter value</a> for <code>sh:uniqueMembers</code>.
+							Each <a>value node</a> <code>v</code> must be a <a href="#syntax-rule-SHACL-list">SHACL list</a> - if <code>v</code> is not a SHACL list there is a <a>validation result</a>.
+							If <code>$uniqueMembers</code> is <code>true</code> and the list <code>v</code> has duplicate members,
+							there is a <a>validation result</a>. There must be at most one <a>validation result</a> for each list <code>v</code>.
```
The sentence "There must be at most one validation result..." should be removed because it is redundant and may not apply in cases where the same value node is reached by different, violated constraints.

jeswr replied:
Resolved in https://github.com/w3c/data-shapes/commit/ca91c2c71ce6526995d39a5ba2eca744ba2d9d39

HolgerKnublauch commented on shacl12-test-suite/tests/core/node/uniqueMembers-001.ttl:
```diff
@@ -0,0 +1,86 @@
+      sh:result [
+          rdf:type sh:ValidationResult ;
+          sh:focusNode _:b1 ;
+          sh:resultMessage "Value does not conform to Shape ex:ListShape. See details for more information." ;
```
Wrong message - no ex:ListShape here.

jeswr replied:
I think this is the right message `ex:ListShape` is the NodeShape defined at the top of this file which requires unique memebrs

HolgerKnublauch commented on shacl12-vocabularies/shacl.ttl:
```diff
@@ -907,13 +907,76 @@ sh:MemberShapeConstraintComponent-memberShape
+	sh:parameter sh:UniqueMembersConstraintComponent-uniqueMembers ;
+	rdfs:isDefinedBy sh: .
+
+sh:MemberShapeConstraintComponent-uniqueMembers
```
Wrong URI - needs to be UniqueMembers...

jeswr replied:
```suggestion
sh:UniqueMembersConstraintComponent-uniqueMembers
```

HolgerKnublauch commented on shacl12-test-suite/tests/core/node/maxListLength-001.ttl:
```diff
@@ -0,0 +1,60 @@
+  rdf:type sh:NodeShape ;
+  sh:maxListLength 2 ;
+  # Satisfies all constraints
+  sh:targetNode ( 1 2 ), rdf:nil ;
```
So my initial comment (that I later deleted) turned out to be correct. This doesn't work. When you use a blank node such as ( 1 2) as targetNode, it will be evaluated as a node expression, which means it will try to walk the rdf:List and then use each list member as target node. To work around this for the test cases, I think all lists need to be URI nodes.

jeswr replied:
Resolved in https://github.com/w3c/data-shapes/pull/416/commits/5c72e8eb8a891b008e8c99c54dbdac60407a51ec

HolgerKnublauch reviewed (changes requested):
I am in the middle of implementing all this in Java and found various issues.

jeswr committed:
chore: rename `uniqList` ->` uniqueMembers` in shacl shacl

jeswr committed:
chore: fix reference to `UniqueMembersConstraintComponent` in vocab

jeswr committed:
chore: remove blank node targetNodes for lists

jeswr committed:
fix: make invalid SHACL lists a top-level violation

jeswr commented:
I have addressed the issues found - is there an(other?) open source implementation that'd be useful for me to implement the changes against (@bergos' JS shacl-engine perhaps) in case there are any more errors to catch.

P.S. I am conferencing for the next 3 days so will be slow to respond

HolgerKnublauch commented on shacl12-test-suite/tests/core/property/maxListLength-001.ttl:
```diff
@@ -0,0 +1,85 @@
+          sh:resultPath ex:hobbies ;
+          sh:resultSeverity sh:Violation ;
+          sh:sourceConstraintComponent sh:MaxListLengthConstraintComponent ;
+          sh:sourceShape ex:PersonShape ;
```
The sourceShape must be the property shape, not the node shape. To implement this it would be best to use URIs for the property shapes. Same issue below and in the minListLength and uniqueMembers tests.

jeswr replied:
Resolved in https://github.com/w3c/data-shapes/pull/416/commits/99ca14f632dba3ca9f623d708864dd8071f375b3

HolgerKnublauch commented on shacl12-vocabularies/shacl.ttl:
```diff
@@ -907,13 +907,76 @@ sh:MemberShapeConstraintComponent-memberShape
+	a sh:Parameter ;
+	sh:path sh:minListLength ;
+	sh:datatype xsd:integer ;
+	sh:maxCount 1 ;
```
Unnecessary: parameters are always maxCount 1

jeswr replied:
```suggestion
```

HolgerKnublauch commented on shacl12-vocabularies/shacl.ttl:
```diff
@@ -907,13 +907,76 @@ sh:MemberShapeConstraintComponent-memberShape
+	a sh:Parameter ;
+	sh:path sh:maxListLength ;
+	sh:datatype xsd:integer ;
+	sh:maxCount 1 ;
```
Unnecessary: parameters are always maxCount 1

jeswr replied:
```suggestion
```

HolgerKnublauch commented on shacl12-vocabularies/shacl.ttl:
```diff
@@ -907,13 +907,76 @@ sh:MemberShapeConstraintComponent-memberShape
+	a sh:Parameter ;
+	sh:path sh:uniqueMembers ;
+	sh:datatype xsd:boolean ;
+	sh:maxCount 1 ;
```
Unnecessary: parameters are always maxCount 1

jeswr replied:
```suggestion
```

HolgerKnublauch commented on shacl12-vocabularies/shacl.ttl:
```diff
@@ -907,13 +907,76 @@ sh:MemberShapeConstraintComponent-memberShape
+	a sh:ConstraintComponent ;
+	rdfs:label "Maximum list length constraint component"@en ;
+	rdfs:comment "Specifies the maximum length of a SHACL list. A violation is reported if the value is not a valid SHACL list."@en ;
+	sh:parameter sh:MaxListLengthConstraintComponent-minListLength ;
```
Typo: minListLength -> maxListLength

jeswr replied:
```suggestion
	sh:parameter sh:MaxListLengthConstraintComponent-maxListLength ;
```

HolgerKnublauch commented on shacl12-test-suite/tests/core/node/maxListLength-001.ttl:
```diff
@@ -0,0 +1,64 @@
+      sh:conforms "false"^^xsd:boolean ;
+      sh:result [
+          rdf:type sh:ValidationResult ;
+          sh:focusNode rdf:nil ;
```
sh:focusNode and also the sh:value must be ex:list1

jeswr replied:
```suggestion
          sh:focusNode ex:list1 ;
```

HolgerKnublauch commented on shacl12-test-suite/tests/core/node/maxListLength-001.ttl:
```diff
@@ -0,0 +1,64 @@
+      sh:result [
+          rdf:type sh:ValidationResult ;
+          sh:focusNode rdf:nil ;
+          sh:resultMessage "List has length 3, but maximum allowed length is 2." ;
```
Test cases should not include the sh:resultMessage because that is up to the implementation, and locale-specific. Right now, my test engine would fail on this, so I guess others may break too.

jeswr commented on shacl12-test-suite/tests/core/node/maxListLength-001.ttl:
```diff
@@ -0,0 +1,64 @@
+          sh:resultMessage "List has length 3, but maximum allowed length is 2." ;
+          sh:resultSeverity sh:Violation ;
+          sh:sourceConstraintComponent sh:MaxListLengthConstraintComponent ;
+          sh:sourceShape ex:ListShape ;
```
```suggestion
          sh:sourceShape ex:ListShape ;
          sh:value ex:list1 ;
```

jeswr committed:
fix: Use correct focusNode and value for maxListLength test

jeswr committed:
fix: remove unnecessary maxCount and fix maxListLength name error in vocab

jeswr committed:
chore: remove result messages from tests

jeswr committed:
chore: make sourceShape a property shape where applicable

jeswr commented on shacl12-vocabularies/shacl-shacl.ttl:
```diff
@@ -74,7 +74,8 @@ shsh:ShapeShape
 		sh:not, sh:or, sh:pattern, sh:property, sh:qualifiedMaxCount, sh:qualifiedMinCount, sh:qualifiedValueShape,
-		sh:qualifiedValueShape, sh:qualifiedValueShapesDisjoint, sh:qualifiedValueShapesDisjoint, sh:uniqueLang, sh:xone ;
+		sh:qualifiedValueShape, sh:qualifiedValueShapesDisjoint, sh:qualifiedValueShapesDisjoint, sh:uniqueLang, sh:xone ,
+		sh:minListLength, sh:maxListLength, sh:uniqList ;
```
```suggestion
		sh:minListLength, sh:maxListLength, sh:uniqueMembers ;
```

jeswr commented on shacl12-vocabularies/shacl-shacl.ttl:
```diff
@@ -226,7 +227,24 @@ shsh:ShapeShape
+	] ;
+	sh:property [
+		sh:path sh:uniqueMembers ;
+		sh:datatype xsd:boolean ;       # uniqList-datatype
+		sh:maxCount 1 ;                 # uniqList-maxCount
```
```suggestion
		sh:datatype xsd:boolean ;       # uniqueMembers-datatype
		sh:maxCount 1 ;                        # uniqueMembers-maxCount
```

jeswr committed:
chore: remove lingering `uniqList` references

HolgerKnublauch committed:
Fixed membershape test case

HolgerKnublauch committed:
Fixed missing sh:value at minListLength test

HolgerKnublauch committed:
Fixed missing sh:value at uniqueMembers test case

jeswr commented:
Thanks for the fixes @HolgerKnublauch

In case needed as future provenance for others: note that https://github.com/w3c/data-shapes/pull/416/commits/58daff683e4d0c3c200018051082235fac217154 makes it such that if a list is malformed - the `sh:value` should always be the head of the list - _not_ the first malformed node.
